src/emulation/promises.ts

```diff
@@ -891,22 +891,32 @@ export function watch(filename: fs.PathLike, options?: fs.WatchOptions | string)
 export function watch<T extends string | Buffer>(filename: fs.PathLike, options: fs.WatchOptions | string = {}): AsyncIterable<promises.FileChangeInfo<T>> {
 	return {
 		[Symbol.asyncIterator](): AsyncIterator<promises.FileChangeInfo<T>> {
-			const watcher = new FSWatcher<T>(filename.toString(), typeof options != 'string' ? options : { encoding: options as BufferEncoding | 'buffer' });
-
-			function withDone(done: boolean) {
-				return function () {
-					const event = Promise.withResolvers<IteratorResult<promises.FileChangeInfo<T>>>();
-					watcher.on('change', (eventType, filename) => {
-						event.resolve({ value: { eventType, filename }, done });
-					});
-					return event.promise;
-				};
+			const watcher = new FSWatcher<T>(filename.toString(), typeof options !== 'string' ? options : { encoding: options as BufferEncoding | 'buffer' });
+
+			// A queue to hold change events, since we need to resolve them in the async iterator
+			const eventQueue: { resolve: (value: IteratorResult<promises.FileChangeInfo<T>>) => void }[] = [];
```

Comment:
Please change the `eventQueue` to be `(value: IteratorResult<promises.FileChangeInfo<T>> => void)[]` (that is, not inside an object).

This would make the iteration simpler:
```ts
for (const resolve of eventQueue) {
	resolve({ value: null, done: true });
}
```

And finally when pushing it onto the event queue:
```ts
eventQueue.push(resolve);
```

Reply:
Ok, but the shift will be a bit tricky, ok?

eventQueue.shift?.()?.({ value: { eventType, filename }, done: false });

Reply:
The first optional chain is unnecessary